From 2f2652ff988b84db627909eac1e6a4ef644144a2 Mon Sep 17 00:00:00 2001
From: Emil Abildgaard Svendsen <EMAS@bang-olufsen.dk>
Date: Thu, 9 Mar 2023 07:54:32 +0100
Subject: [PATCH 2/2] ASoC: hdmi-codec: only startup/shutdown on supported
 streams

commit e041a2a550582106cba6a7c862c90dfc2ad14492 upstream.

Currently only one stream is supported. This isn't usally a problem
until you have a multi codec audio card. Because the audio card will run
startup and shutdown on both capture and playback streams. So if your
hdmi-codec only support either playback or capture. Then ALSA can't open
for playback and capture.

This patch will ignore if startup and shutdown are called with a non
supported stream. Thus, allowing an audio card like this:

           +-+
 cpu1 <--@-| |-> codec1 (HDMI-CODEC)
           | |<- codec2 (NOT HDMI-CODEC)
           +-+

Signed-off-by: Emil Svendsen <emas@bang-olufsen.dk>
Link: https://lore.kernel.org/r/20230309065432.4150700-2-emas@bang-olufsen.dk
Signed-off-by: Mark Brown <broonie@kernel.org>
[smazurov@gmail.com: rebased onto Armbian linux-rockchip rk-6.1-rkr5.1 which
 carries an additional tx_dlp early-return in both hdmi_codec_startup and
 hdmi_codec_shutdown (introduced by commit e22dfe2e, "ASoC: hdmi-codec: Add
 support for HDMI-TX DLP"). The new has_playback / has_capture guard from
 this upstream commit is added immediately AFTER the tx_dlp guard, so both
 mechanisms compose: tx_dlp short-circuits HDMI-TX DLP capture paths,
 has_capture/has_playback short-circuits unsupported-direction calls on
 codecs that advertise the no_i2s_* flags from the companion patch.

 Companion to the no_i2s_capture restoration. Both are needed together;
 without this patch ALSA can still hit "Only one simultaneous stream
 supported!" on multi-codec cards where the HDMI-codec advertises only
 one direction.]
Tested-by: Stepan Mazurov <smazurov@gmail.com>
Signed-off-by: Stepan Mazurov <smazurov@gmail.com>
---
 sound/soc/codecs/hdmi-codec.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/sound/soc/codecs/hdmi-codec.c b/sound/soc/codecs/hdmi-codec.c
index e86d12e20..4da7376e8 100644
--- a/sound/soc/codecs/hdmi-codec.c
+++ b/sound/soc/codecs/hdmi-codec.c
@@ -459,11 +459,16 @@ static int hdmi_codec_startup(struct snd_pcm_substream *substream,
 {
 	struct hdmi_codec_priv *hcp = snd_soc_dai_get_drvdata(dai);
 	bool tx = substream->stream == SNDRV_PCM_STREAM_PLAYBACK;
+	bool has_capture = !hcp->hcd.no_i2s_capture;
+	bool has_playback = !hcp->hcd.no_i2s_playback;
 	int ret = 0;
 
 	if (hcp->tx_dlp && substream->stream != SNDRV_PCM_STREAM_PLAYBACK)
 		return 0;
 
+	if (!((has_playback && tx) || (has_capture && !tx)))
+		return 0;
+
 	mutex_lock(&hcp->lock);
 	if (hcp->busy) {
 		dev_err(dai->dev, "Only one simultaneous stream supported!\n");
@@ -503,10 +508,16 @@ static void hdmi_codec_shutdown(struct snd_pcm_substream *substream,
 				struct snd_soc_dai *dai)
 {
 	struct hdmi_codec_priv *hcp = snd_soc_dai_get_drvdata(dai);
+	bool tx = substream->stream == SNDRV_PCM_STREAM_PLAYBACK;
+	bool has_capture = !hcp->hcd.no_i2s_capture;
+	bool has_playback = !hcp->hcd.no_i2s_playback;
 
 	if (hcp->tx_dlp && substream->stream != SNDRV_PCM_STREAM_PLAYBACK)
 		return;
 
+	if (!((has_playback && tx) || (has_capture && !tx)))
+		return;
+
 	hcp->chmap_idx = HDMI_CODEC_CHMAP_IDX_UNKNOWN;
 	hcp->hcd.ops->audio_shutdown(dai->dev->parent, hcp->hcd.data);
 
